pkg/kubelet/certificate/certificate_manager.go

```diff
@@ -0,0 +1,371 @@
+			keys[withoutExt(keyfile.Name())] = false
+		}
+	}
+	certfiles, err := ioutil.ReadDir(m.certDirectory)
+	if err != nil {
+		return err
+	}
+	for _, certfile := range certfiles {
+		if filepath.Ext(certfile.Name()) == certExtension {
+			if _, ok := keys[withoutExt(certfile.Name())]; ok {
+				keys[withoutExt(certfile.Name())] = true
+			} else {
+				certPath := filepath.Join(m.certDirectory, certfile.Name())
+				if err := os.Remove(certPath); err != nil {
+					glog.Errorf("The certificate %q doesn't have a matching key but can't be deleted: %v", certPath, err)
+				}
+			}
+		}
+	}
+	for filename, certFound := range keys {
+		if !certFound {
+			keyPath := filepath.Join(m.keyDirectory, filename+keyExtension)
+			if err := os.Remove(keyPath); err != nil {
+				glog.Errorf("The certificate %q doesn't have a matching key but can't be deleted: %v", keyPath, err)
+			}
+		}
+	}
+	return nil
+}
+
+func fileExists(file string) bool {
+	_, err := os.Stat(file)
+	return err == nil || os.IsExist(err)
+}
+
+// withoutExt returns the given filename after removing the extension. The
+// extension to remove will be the result of filepath.Ext().
+func withoutExt(filename string) string {
+	return strings.TrimSuffix(filename, filepath.Ext(filename))
+}
+
+// fixSymLink moves the given file to another location and replaces it with a
+// symlink to the new location.
+func fixSymlink(filename string) error {
+	// Check if the given file is already a symlink.
+	if fi, err := os.Lstat(filename); err != nil {
+		return err
+	} else if fi.Mode()&os.ModeSymlink == os.ModeSymlink {
+		return nil
+	}
+
+	newFilename := withoutExt(filename) + ".orig" + filepath.Ext(filename)
+
+	// Move the file to the backup location.
+	if err := os.Rename(filename, newFilename); err != nil {
+		return fmt.Errorf("unable to rename %q to %q: %v", filename, newFilename, err)
+	}
+	if err := os.Symlink(newFilename, filename); err != nil {
+		return fmt.Errorf("unable to create a symlink from %q to %q: %v", newFilename, filename, err)
+	}
+	return nil
+}
+
+func filename(qualifier, extension string) string {
+	return pairNamePrefix + "-" + qualifier + extension
+}
+
+// shouldRotate looks at how close the current certificate is to expiring and
+// decides if it is time to rotate or not.
+func (m *manager) shouldRotate() bool {
+	notAfter := m.cert.Leaf.NotAfter
+	total := notAfter.Sub(m.cert.Leaf.NotBefore)
+	remaining := notAfter.Sub(time.Now())
+	return (remaining * 100 / total) < shouldRotatePercent
+}
+
+func (m *manager) rotateCerts() error {
+	if !m.shouldRotate() {
+		return nil
+	}
+
+	ts := time.Now().Format("2006-01-02-15-04-05")
+	keyFilename := filename(ts, keyExtension)
+	certFilename := filename(ts, certExtension)
+
+	success := false
+
+	// Get the private key.
+	keyPath := filepath.Join(m.keyDirectory, keyFilename)
+	keyData, generatedKeyFile, err := certutil.LoadOrGenerateKeyFile(keyPath)
+	if err != nil {
+		return err
+	}
+	if generatedKeyFile {
```

Comment:
When would this ever not generate the keyfile? I almost feel like it should be an error if it's not generated... Or maybe refactor `LoadOrGenerateKeyFile` so the generate portion can be called directly.